File: density/src/density_pce_aware.py
import cv2
import numpy as np
from ultralytics import YOLO
import supervision as sv

# ==========================================
# CẤU HÌNH THÔNG SỐ
# ==========================================
MODEL_PATH = "best.onnx"
VIDEO_PATH = "test_video.mp4"
OUTPUT_PATH = "output_density_pce.mp4"

# # Tọa độ ROI (Kế thừa từ bản chuẩn của bạn)
# ROI_VERTICES = np.array([
#     [600, 400],   
#     [1300, 400],  
#     [2000, 950],  
#     [200, 975]    
# ], dtype=np.float32) # Giữ float32 nếu sau này bạn muốn tích hợp chung với BEV, còn ở hàm này int32 hay float32 đều chạy được.


# Tọa độ ROI (Kế thừa từ bản chuẩn của bạn)
ROI_VERTICES = np.array([
    [750, 400],   # Điểm 1: Góc trái trên
    [980, 400],   # Điểm 2: Góc phải trên
    [1250, 1050],  # Điểm 3: Góc phải dưới (mở rộng ra do góc phối cảnh camera)
    [200, 1000]    # Điểm 4: Góc trái dưới
], dtype=np.float32) # Giữ float32 nếu sau này bạn muốn tích hợp chung với BEV, còn ở hàm này int32 hay float32 đều chạy được.

# # Tọa độ ROI (Kế thừa từ bản chuẩn của bạn)
# ROI_VERTICES = np.array([
#     [784, 359],   # Điểm 1: Góc trái trên
#     [968, 359],   # Điểm 2: Góc phải trên
#     [1155, 748],  # Điểm 3: Góc phải dưới (mở rộng ra do góc phối cảnh camera)
#     [359, 748]    # Điểm 4: Góc trái dưới
# ], dtype=np.float32) # Giữ float32 nếu sau này bạn muốn tích hợp chung với BEV, còn ở hàm này int32 hay float32 đều chạy được.

# ==========================================
# HÀM 3: BỘ TỪ ĐIỂN HỆ SỐ QUY ĐỔI PCE
# ==========================================
# Lưu ý: Các key trong từ điển này PHẢI KHỚP CHÍNH XÁC 100% với 
# mảng names trong file data.yaml của bạn (bus, car, motor, truck)
PCE_WEIGHTS = {
    "motor": 0.5,
    "car": 1.0,
    "bus": 2.5,
    "truck": 2.5
}

# ==========================================
# NGƯỠNG ĐÁNH GIÁ ÙN TẮC (CONGESTION THRESHOLDS)
# ==========================================
# Dựa trên tổng điểm PCE trong vùng ROI để kết luận trạng thái.
# Các con số này cần tinh chỉnh dựa trên diện tích thực tế của ROI.
THRESHOLD_HEAVY = 15.0  
THRESHOLD_JAM = 25.0    

def calculate_pce_density(detections_in_roi, class_names_dict):
    """
    Hàm tính toán tổng mật độ PCE của các xe trong vùng ROI.
    
    Tham số:
        detections_in_roi: Dữ liệu xe đã được lọc (chỉ lấy xe nằm trong ROI).
        class_names_dict: Từ điển map ID class sang tên class của YOLO (VD: {0: 'bus', 1: 'car'...}).
        
    Trả về:
        float: Tổng điểm PCE.
        list: Danh sách nhãn dán chi tiết cho từng xe (để hiển thị).
    """
    total_pce = 0.0
    labels = []
    
    # Lặp qua từng xe trong ROI
    for class_id, tracker_id, conf in zip(detections_in_roi.class_id, detections_in_roi.tracker_id, detections_in_roi.confidence):
        # Lấy tên class (VD: 'motor', 'car')
        class_name = class_names_dict[class_id]
        
        # Tra bảng lấy hệ số PCE. Nếu model nhận diện ra class lạ, mặc định gán = 1.0
        pce_val = PCE_WEIGHTS.get(class_name, 1.0)
        
        # Cộng dồn vào tổng điểm PCE của khung hình
        total_pce += pce_val
        
        # Tạo nhãn hiển thị cho xe: "#ID Tên_xe (PCE: Điểm) Độ_tự_tin"
        labels.append(f"#{tracker_id} {class_name} (PCE:{pce_val}) {conf:.2f}")
        
    return total_pce, labels


def main():
    print("[INFO] Khởi động Hàm 3: Ước lượng Mật độ PCE (PCE-Aware)...")
    model = YOLO(MODEL_PATH)
    
    video_info = sv.VideoInfo.from_video_path(VIDEO_PATH)
    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
    out = cv2.VideoWriter(OUTPUT_PATH, fourcc, video_info.fps, (video_info.width, video_info.height))

    # Cấu hình công cụ vẽ: Vẽ viền mỏng hơn chút để dễ nhìn xe khi đông
    box_annotator = sv.BoxAnnotator(thickness=2)
    label_annotator = sv.LabelAnnotator(text_thickness=2, text_scale=0.6, text_color=sv.Color.BLACK)

    # Đa giác ROI
    polygon_zone = sv.PolygonZone(polygon=ROI_VERTICES.astype(np.int32))
    polygon_annotator = sv.PolygonZoneAnnotator(zone=polygon_zone, color=sv.Color.RED, thickness=3)

    cap = cv2.VideoCapture(VIDEO_PATH)

    cv2.namedWindow("PCE Density Estimation", cv2.WINDOW_NORMAL)
    cv2.resizeWindow("PCE Density Estimation", 1280, 720) 

    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        results = model.track(frame, persist=True, tracker="bytetrack.yaml", conf=0.3, verbose=False)[0]

        total_pce = 0.0

        if results.boxes.id is not None:
            detections = sv.Detections.from_ultralytics(results)
            
            # 1. Dùng Mask để lọc xe trong vùng ROI trước
            mask = polygon_zone.trigger(detections=detections)
            detections_in_roi = detections[mask]

            # 2. Gọi Hàm 3 tính điểm PCE
            # results.names là một dictionary map từ class_id sang class_name của model YOLO
            total_pce, custom_labels = calculate_pce_density(detections_in_roi, results.names)

            # 3. Vẽ Box và Nhãn dán PCE lên các xe
            frame = box_annotator.annotate(scene=frame, detections=detections_in_roi)
            frame = label_annotator.annotate(scene=frame, detections=detections_in_roi, labels=custom_labels)

        # ==========================================
        # # ĐÁNH GIÁ TRẠNG THÁI ÙN TẮC VÀ ĐỔI MÀU ROI
        # # ==========================================
        # if total_pce >= THRESHOLD_JAM:
        #     status_text = "TRAFFIC JAM (Kẹt xe cứng)"
        #     status_color = (0, 0, 255) # Đỏ BGR
        #     poly_color = sv.Color.RED
        # elif total_pce >= THRESHOLD_HEAVY:
        #     status_text = "HEAVY (Ùn ứ)"
        #     status_color = (0, 165, 255) # Cam BGR
        #     poly_color = sv.Color(255, 165, 0)
        # else:
        #     status_text = "NORMAL (Thông thoáng)"
        #     status_color = (0, 255, 0) # Xanh lá BGR
        #     poly_color = sv.Color.GREEN

        # # Đổi màu đa giác theo trạng thái
        # polygon_annotator.color = poly_color
        # frame = polygon_annotator.annotate(scene=frame)

        # # ==========================================
        # # VẼ DASHBOARD HIỂN THỊ LÊN MÀN HÌNH
        # # ==========================================
        # cv2.rectangle(frame, (20, 20), (700, 180), (0, 0, 0), -1) # Bảng đen mờ
        
        # cv2.putText(frame, f"Absolute Count (N): {len(detections_in_roi if results.boxes.id is not None else [])}", 
        #             (40, 60), cv2.FONT_HERSHEY_SIMPLEX, 1.2, (255, 255, 255), 3)
        
        # cv2.putText(frame, f"Total PCE Point: {total_pce:.1f}", 
        #             (40, 110), cv2.FONT_HERSHEY_SIMPLEX, 1.2, (0, 255, 255), 3)
        
        # cv2.putText(frame, f"Status: {status_text}", 
        #             (40, 160), cv2.FONT_HERSHEY_SIMPLEX, 1.2, status_color, 3)

        # out.write(frame)
        # cv2.imshow("PCE Density Estimation", frame)

        # ==========================================
        # TOÁN HỌC: TÍNH MẬT ĐỘ PCE THỰC TẾ (k_pce = Tổng PCE / L)
        # ==========================================
        # Khai báo chiều dài đoạn đường thực tế trong khung ROI (Ví dụ: 15 mét)
        # Thông số này cần được "hiệu chuẩn" (calibrate) một lần khi lắp camera thực tế
        # ROAD_LENGTH_KM = 0.015 
        ROAD_LENGTH_KM = 0.1;
        pce_density = 0.0
        if ROAD_LENGTH_KM > 0:
            pce_density = total_pce / ROAD_LENGTH_KM

        # ==========================================
        # ĐÁNH GIÁ TRẠNG THÁI DỰA TRÊN MẬT ĐỘ PCE/KM (Chuẩn quốc tế)
        # ==========================================
        # Thay vì dùng Threshold cố định cho khung ảnh, ta dùng Threshold cho 1km
        # Các ngưỡng dưới đây có thể tùy chỉnh:
        # > 1500 PCE/km: Kẹt xe nặng
        # > 800 PCE/km: Ùn ứ
        if pce_density >= 1500.0:
            status_text = "TRAFFIC JAM (Ket xe)"
            status_color = (0, 0, 255) # Đỏ
            polygon_annotator.color = sv.Color.RED
        elif pce_density >= 800.0:
            status_text = "HEAVY (Un u)"
            status_color = (0, 165, 255) # Cam
            polygon_annotator.color = sv.Color(255, 165, 0)
        else:
            status_text = "NORMAL (Thong thoang)"
            status_color = (0, 255, 0) # Xanh lá
            polygon_annotator.color = sv.Color.GREEN

        # Vẽ đa giác đã đổi màu
        frame = polygon_annotator.annotate(scene=frame)

        # ==========================================
        # VẼ DASHBOARD HIỂN THỊ LÊN MÀN HÌNH
        # ==========================================
        cv2.rectangle(frame, (20, 20), (750, 220), (0, 0, 0), -1) 
        
        cv2.putText(frame, f"Absolute Count: {len(detections_in_roi if results.boxes.id is not None else [])} vehicles", 
                    (40, 60), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (255, 255, 255), 2)
        
        cv2.putText(frame, f"Total PCE Points: {total_pce:.1f} PCE", 
                    (40, 110), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 255), 2)
        
        cv2.putText(frame, f"Density (k): {pce_density:.1f} PCE/km", 
                    (40, 160), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (255, 0, 255), 2)

        cv2.putText(frame, f"Status: {status_text}", 
                    (40, 210), cv2.FONT_HERSHEY_SIMPLEX, 1.2, status_color, 3)

        out.write(frame)
        cv2.imshow("PCE Density Estimation", frame)

        # ... (Phần còn lại giữ nguyên) ...
        

        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    cap.release()
    out.release()
    cv2.destroyAllWindows()
    print("[SUCCESS] Đã hoàn thành Hàm 3: Ước lượng PCE.")

if __name__ == "__main__":
    main()

File: density/src/cpp/density_pce_aware.cpp
// =============================================================================
// density_pce_aware.cpp
// Method 4: PCE-Weighted Density  k_pce = sum(PCE_i) / L
//
// Python equivalent: density_pce_aware.py
// Algorithm:
//   - For each vehicle inside ROI, look up its class name in PCE_WEIGHTS.
//   - Sum all PCE values → total_pce.
//   - k_pce = total_pce / road_length_km  (PCE/km)
//   - Color-code ROI polygon and dashboard by congestion level.
//
// PCE (Passenger Car Equivalent) converts each vehicle type to an equivalent
// number of passenger cars, enabling fair multi-class density comparison.
// =============================================================================
#include "density_common.h"
#include <iostream>
#include <unordered_map>

// ==========================================
// CONFIGURATION
// ==========================================
static const std::string MODEL_PATH  = "best.onnx";
static const std::string VIDEO_PATH  = "test_video.mp4";
static const std::string OUTPUT_PATH = "output_density_pce.mp4";

// Class names must match data.yaml order
static const std::vector<std::string> CLASS_NAMES = {"bus", "car", "motor", "truck"};

// ROI polygon
static const std::vector<cv::Point> ROI_VERTICES = {
    {750,  400},
    {980,  400},
    {1250, 1050},
    {200,  1000}
};

// Road length inside ROI (km) — calibrate once per camera installation
static const double ROAD_LENGTH_KM = 0.1;

// ==========================================
// PCE WEIGHT TABLE
// Keys must match CLASS_NAMES exactly.
// ==========================================
static const std::unordered_map<std::string, double> PCE_WEIGHTS = {
    {"motor", 0.5},
    {"car",   1.0},
    {"bus",   2.5},
    {"truck", 2.5}
};

// ==========================================
// CONGESTION THRESHOLDS (PCE/km — international standard)
// ==========================================
static const double THRESHOLD_HEAVY = 800.0;   // > 800  PCE/km → ùn ứ       (heavy)
static const double THRESHOLD_JAM   = 1500.0;  // > 1500 PCE/km → kẹt xe     (jam)

// ==========================================
// FUNCTION 4: PCE density
// Replaces: calculate_pce_density() in Python
// Returns: {total_pce, labels}
// ==========================================
static std::pair<double, std::vector<std::string>>
calculate_pce_density(const std::vector<Detection>& dets_in_roi)
{
    double total_pce = 0.0;
    std::vector<std::string> labels;

    for (const auto& d : dets_in_roi) {
        const std::string& name = CLASS_NAMES[d.class_id];

        // Look up PCE weight; default = 1.0 for unknown classes
        double pce_val = 1.0;
        auto it = PCE_WEIGHTS.find(name);
        if (it != PCE_WEIGHTS.end()) pce_val = it->second;

        total_pce += pce_val;

        char buf[64];
        std::snprintf(buf, sizeof(buf), "#%d %s (PCE:%.1f) %.2f",
                      d.tracker_id, name.c_str(), pce_val, d.confidence);
        labels.push_back(buf);
    }

    return {total_pce, labels};
}

// ==========================================
// Helper: congestion status from k_pce
// ==========================================
struct Status {
    std::string text;
    cv::Scalar  color_bgr;
    cv::Scalar  roi_color;
};

static Status get_status(double k_pce) {
    if (k_pce >= THRESHOLD_JAM)
        return {"TRAFFIC JAM (Ket xe)",       {0,0,255},   {0,0,255}};
    if (k_pce >= THRESHOLD_HEAVY)
        return {"HEAVY (Un u)",               {0,165,255}, {0,165,255}};
    return    {"NORMAL (Thong thoang)",       {0,255,0},   {0,255,0}};
}

int main() {
    std::cout << "[INFO] Method 4: PCE-Aware Density (k_pce = sum(PCE) / L)\n";

    YoloDetector detector(MODEL_PATH, CLASS_NAMES);
    SimpleTracker tracker;
    PolygonZone   zone(ROI_VERTICES);

    cv::VideoCapture cap(VIDEO_PATH);
    if (!cap.isOpened()) { std::cerr << "[ERROR] Cannot open video\n"; return 1; }

    int    fw  = (int)cap.get(cv::CAP_PROP_FRAME_WIDTH);
    int    fh  = (int)cap.get(cv::CAP_PROP_FRAME_HEIGHT);
    double fps = cap.get(cv::CAP_PROP_FPS);

    cv::VideoWriter writer(OUTPUT_PATH,
                           cv::VideoWriter::fourcc('m','p','4','v'),
                           fps, {fw, fh});

    cv::namedWindow("PCE Density Estimation", cv::WINDOW_NORMAL);
    cv::resizeWindow("PCE Density Estimation", 1280, 720);

    cv::Mat frame;
    while (cap.read(frame)) {
        auto dets = detector.detect(frame);
        tracker.update(dets);

        // ── Filter: only vehicles inside ROI ─────────────────────────────
        auto mask = zone.trigger(dets);
        std::vector<Detection> dets_in_roi;
        for (size_t i = 0; i < dets.size(); i++)
            if (mask[i]) dets_in_roi.push_back(dets[i]);

        // ── CORE ALGORITHM: PCE density ───────────────────────────────────
        auto [total_pce, labels] = calculate_pce_density(dets_in_roi);

        // k_pce = total PCE points / road length (PCE/km)
        double k_pce = (ROAD_LENGTH_KM > 0) ? total_pce / ROAD_LENGTH_KM : 0.0;

        // ── Congestion status ─────────────────────────────────────────────
        Status status = get_status(k_pce);

        // ── Draw: ROI polygon (color changes with congestion level) ───────
        std::vector<std::vector<cv::Point>> poly_pts = {ROI_VERTICES};
        cv::polylines(frame, poly_pts, true, status.roi_color, 3);

        // ── Draw: detections with PCE labels ──────────────────────────────
        for (size_t i = 0; i < dets_in_roi.size(); i++) {
            cv::Rect r(dets_in_roi[i].bbox);
            cv::rectangle(frame, r, {0, 255, 0}, 2);
            cv::putText(frame, labels[i], {r.x, r.y - 5},
                        cv::FONT_HERSHEY_SIMPLEX, 0.45, {0, 255, 0}, 1, cv::LINE_AA);
        }

        // ── HUD ───────────────────────────────────────────────────────────
        char buf_count[64], buf_pce[64], buf_k[64];
        std::snprintf(buf_count, sizeof(buf_count),
                      "Absolute Count: %d vehicles", (int)dets_in_roi.size());
        std::snprintf(buf_pce,   sizeof(buf_pce),
                      "Total PCE Points: %.1f PCE",  total_pce);
        std::snprintf(buf_k,     sizeof(buf_k),
                      "Density (k): %.1f PCE/km",    k_pce);

        draw_info_box(frame, {
            {buf_count,                       {255, 255, 255}},
            {buf_pce,                         {0,   255, 255}},
            {buf_k,                           {255,   0, 255}},
            {"Status: " + status.text,        status.color_bgr}
        });

        writer.write(frame);
        cv::imshow("PCE Density Estimation", frame);
        if ((cv::waitKey(1) & 0xFF) == 'q') break;
    }

    cap.release();
    writer.release();
    cv::destroyAllWindows();
    std::cout << "[SUCCESS] Method 4 complete.\n";
    return 0;
}
